File: typescript/src/day1.ts
import { readFileSync } from "fs"

const input = readFileSync("../inputs/day1.txt", "utf-8")

export const parseInputToArrays = (input: string): number[][] => {
    const lines = input.split("\r\n")

    const firstArr: number[] = []
    const secondArr: number[] = []

    lines.forEach((line) => {
        // Split the line at whitespace to get the two values
        const [first, second] = line.split(/\s+/)
        // Push them to the output arrays
        firstArr.push(Number(first))
        secondArr.push(Number(second))
    })

    return [firstArr, secondArr]
}

export const part1 = (first: number[], second: number[]): number => {
    // Sort both arrays
    first.sort((a, b) => a - b)
    second.sort((a, b) => a - b)

    const output = first.reduce((acc, val, index) => {
        return acc + Math.abs(val - second[index])
    }, 0)

    return output
}

export const part2 = (first: number[], second: number[]): number => {
    let result: number = 0

    first.forEach((value) => {
        const count = second.reduce(
            (acc, val) => (val === value ? acc + 1 : acc),
            0
        )
        result += value * count
    })

    return result
}

const [first, second] = parseInputToArrays(input)

const part1Result = part1(first, second)
const part2Result = part2(first, second)

console.log("The result for day 1 part 1 is ", part1Result)
console.log("The result for day 1 part 2 is ", part2Result)

File: rust/src/bin/day1.rs
const INPUT: &str = include_str!("../../../inputs/day1.txt");

fn main() {
    println!("The result for day 1 part 1 is {}", part1(INPUT));
}

fn part1(input: &str) -> i32 {
    let mut first_vec: Vec<i32> = Vec::new();
    let mut second_vec: Vec<i32> = Vec::new();

    for line in input.lines() {
        // Split the line by the white space in between each number, put each number into the corresponding vector
        let mut split = line.split_whitespace();

        if let (Some(first), Some(second)) = (split.next(), split.next()) {
            first_vec.push(first.parse().unwrap());
            second_vec.push(second.parse().unwrap());
        } else {
            eprintln!("Line format is invalid: {}", line);
        }
    }

    // Sort vectors in descending order
    first_vec.sort_by(|a, b| b.cmp(a));
    second_vec.sort_by(|a, b| b.cmp(a));

    // Iterate through both vectors at once, find the absolute difference between them and sum it all up
    first_vec
        .iter()
        .zip(second_vec.iter())
        .map(|(first, second)| (first - second).abs())
        .sum()
}

#[cfg(test)]
mod tests {
    const TEST_INPUT: &str = include_str!("../../../testInputs/day1.txt");
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(part1(TEST_INPUT), 11);
    }
}
